File: javascript/Sorting/mergeSort.js


function mergeArrays(arr1, arr2) {

    result = []
    i = 0
    j = 0

    while (i < arr1.length && j < arr2.length) {

        if (arr1[i] >= arr2[j]) {
            result.push(arr2[j])
            j++
        }
        else {
            result.push(arr1[i])
            i++
        }
    }

    while (i < arr1.length) {
        result.push(arr1[i])
        i++
    }

    while (j < arr2.length) {
        result.push(arr2[j])
        j++
    }

    return result
}

function mergeSort(arr) {

    if (arr.length <= 1) {
        return arr
    }
    let mid = Math.floor(arr.length / 2)
    let left = mergeSort(arr.slice(0, mid))
    let right = mergeSort(arr.slice(mid))
    results = mergeArrays(left, right)

    return results
}

arr = [97, 2, 1, 34, 9, 74, 4]
console.log(mergeSort(arr))

File: python/Sorting/merge-sort.py
import math

def merge(arr1,arr2):
    leftIdx = 0
    rightIdx=0
    result = []
    while leftIdx < len(arr1) and rightIdx < len(arr2):
        if arr1[leftIdx] > arr2[rightIdx]:
            result.append(arr2[rightIdx])
            rightIdx+=1
        else:
            result.append(arr1[leftIdx])
            leftIdx+=1

    while leftIdx < len(arr1):
            result.append(arr1[leftIdx])
            leftIdx+=1


    while rightIdx < len(arr2):
            result.append(arr2[rightIdx])
            rightIdx+=1
    
    return result

def mergeSort(arr):
     
     if len(arr) <=1: 
          return arr
     
     mid = math.floor(len(arr)/2)

     left = mergeSort(arr[:mid])
     right = mergeSort(arr[mid:])
     return merge(left,right)


arr = [57, 34, 56,1, 32, 17, 99,2, 876, 456]
print(mergeSort(arr))
# Learn how to code mergeSort with the help of Auxilary array so that
# we can reduce the space complexity from O(nlogn) to O(n)